File: Week 1/Union Find/Quick Union/Improvements/Weighted Quick Union/weightedquickunion.py
# Weighted Quick Union (Union Find) is a data structure which improves upon the regular quick union by making sure trees dont get to large by maintaining]
# a seperate array for keeping track of size.
# NOTE: size here refers to the number of objects in the connected component. For example sz[i] means size of connected component with i as root

class WeightedQuickUnionUF:
    def __init__(self, N):
        self.array = [i for i in range(N)]
        self.sz = [1 for i in range(N)]

    def root(self, p):
        while(self.array[p] != p):
            p = self.array[p]
        return p

    def union(self, p, q):
        leftR = self.root(p)
        rightR = self.root(q)
        if(self.sz[leftR] < self.sz[rightR]):
            self.array[leftR] = rightR
            self.sz[rightR] += self.sz[leftR]
        else:
            self.array[rightR] = leftR
            self.sz[leftR] += self.sz[rightR]

    def connected(self, p, q):
        return self.root(p) == self.root(q)

    def display(self):
        for node in range(len(self.array)):
            print(self.array[node], end=" ")
        print()

    def displaySz(self):
        for node in range(len(self.sz)):
            print(self.sz[node], end=" ")
        print()

victim = WeightedQuickUnionUF(10)
victim.display()
victim.displaySz()
victim.union(4, 3)
victim.union(3, 8)
victim.union(6, 5)
victim.union(9, 4)
victim.union(2, 1)
victim.union(5, 0)
victim.union(7, 2)
victim.union(6, 1)
victim.union(7, 3)

victim.display()
victim.displaySz()

File: Week 1/Union Find/Quick Union/Improvements/Weighted Quick Union/weightedquickunion.cpp
#include <iostream>
#include <vector>

using namespace std;

class WeightedQuickUnion
{
private:
    vector<int> array;
    vector<int> sz;
public:
    WeightedQuickUnion(int N) {
        for(int index = 0; index < N; index++) 
            array.push_back(index);
        for(int index = 0; index < N; index++) 
            sz.push_back(1);
    }

    int root(int p) {
        while(array.at(p) != p) 
            p = array.at(p);
        return p;
    }

    void Union(int p, int q) {
        int leftR = root(p);
        int rightR = root(q);
        if(sz[leftR] < sz[rightR]) {
            array[leftR] = rightR;
            sz[rightR] += sz[leftR];
        } else {
            array[rightR] = leftR;
            sz[leftR] += sz[rightR];
        }
    }

    int connected(int p, int q) {
        return root(p) == root(q);
    }

    void display() {
        for(int index = 0; index < array.size(); index++)
            printf("%d ", array[index]);
        printf("\n");
    }

    void displaysz() {
        for(int index = 0; index < array.size(); index++)
            printf("%d ", sz[index]);
        printf("\n");
    }
};

int main()
{
    WeightedQuickUnion victim(10);
    victim.display();
    victim.displaysz();

    victim.Union(4, 3);
    victim.Union(3, 8);
    victim.Union(6, 5);
    victim.Union(9, 4);
    victim.Union(2, 1);
    victim.Union(5, 0);
    victim.Union(7, 2);
    victim.Union(6, 1);
    victim.Union(7, 3);

    victim.display();
    victim.displaysz();
}